### PR TITLE
feat(rust,python): add `include_key` parameter to `partition_by`

### LINKED ISSUE #8808: Add option to remove partitioned column in partition_by
### Problem description

When you use `df.partion_by(on=..., as_dict=True)`, then your dictionary keys hold the same values as the `on` column, and the column becomes redundant information. I find myself often manually having to remove that column:

```python
dfs = df.partition_by(on=key, as_dict=True)

for df in dfs.values():
    df.drop_in_place(key)
```

Having the ability to auto-drop the partitioned column, which is often no longer needed, would make things a bit more ergonomic. A simple `drop_key` parameter, perhaps:

```python
dfs = df.partition_by(on=key, as_dict=True, drop_key=True)
```

### PR DESCRIPTION
**Pull request by mcrumiller**
Resolves #8808.

@ritchie46 I didn't get any feedback on this although it did get a few thumbs up, so I'll need your blessing on this one. Dropping the partitioned columns is something I end up doing all the time when partitioning, as I no longer need them.

```python
df = pl.DataFrame({
    'a': [1, 2, 3, 4, 5],
    'b': [6, 7, 8, 9, 10],
    'c': ['a', 'a', 'a', 'b', 'b']
})

df.partition_by('c', drop_keys=True)
```
```
[shape: (3, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 1   ┆ 6   │
│ 2   ┆ 7   │
│ 3   ┆ 8   │
└─────┴─────┘, shape: (2, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 4   ┆ 9   │
│ 5   ┆ 10  │
└─────┴─────┘]
```

### CODE REVIEW

**Comment by avimallu** [comment]
A very minor suggestion - perhaps consider naming `drop_keys` as `include_key` to be consistent with `rows_by_key`?

**Comment by ritchie46** [comment]
@mcrumiller can you go for @avimallu's suggestion? Then we can merge this. :+1:

**Comment by mcrumiller** [comment]
@ritchie46 done. I switched from `drop_keys=False` default to `include_key=True` to keep the same default functionality.

File: polars/polars-core/src/frame/mod.rs [hunk 1]
```diff
@@ -3346,29 +3347,51 @@ impl DataFrame {
...
+                            } else {
+                                group.drop_many(cols)
+                            }
+                        })
                         .collect())
                 }
                 GroupsProxy::Slice { groups, .. } => Ok(groups
                     .into_par_iter()
                     .map(|[first, len]| self.slice(first as i64, len as usize))
+                    .map(|group| {
```
Comment (ritchie46) [inline polars/polars-core/src/frame/mod.rs:3362]: Can you join the maps into a single closure?

File: polars/polars-core/src/frame/mod.rs [hunk 2]
```diff
@@ -3346,29 +3347,51 @@ impl DataFrame {
...
+                        if include_key {
+                            group
+                        } else {
+                            group.drop_many(cols)
```
Comment (ritchie46) [inline polars/polars-core/src/frame/mod.rs:3366]: Can you use `drop_many_amortized`. That saves an allocation per group. See #9814

Comment (mcrumiller) [reply polars/polars-core/src/frame/mod.rs:3366]: Thanks @ritchie46--I'll do that once that gets merged into main (`drop_many_ amortized` isnt' available yet).

File: polars/polars-core/src/frame/mod.rs (hunk 1 above)
Comment (ritchie46) [reply polars/polars-core/src/frame/mod.rs:3362]: Yes it would. The you also don't need to amortize.

File: polars/polars-core/src/frame/mod.rs [hunk 3]
```diff
@@ -3328,13 +3328,21 @@ impl DataFrame {
...
+        include_key: bool,
     ) -> PolarsResult<Vec<DataFrame>> {
         let groups = if stable {
             self.groupby_stable(cols)?.take_groups()
         } else {
             self.groupby(cols)?.take_groups()
         };
 
+        // drop key columns prior to calculation if requested
+        let df = if !include_key {
```
Comment (ritchie46) [inline polars/polars-core/src/frame/mod.rs:3340]: nit: No need to negate. Simply swap the expressions. That reads a little easier.

Comment (mcrumiller) [reply polars/polars-core/src/frame/mod.rs]: Good catch. fixed.

**Comment by ritchie46** [comment]
Alright, looks good @mcrumiller. Thanks!

**Pull request merged**